### PR TITLE
Include non-convex polygon struct `VPolygonNC` :eight_spoked_asterisk:

### PR DESCRIPTION
**Pull request by mvanzulli**
The polygon interface aim to consider geometric figures like non-convex [polygons in 2D](https://en.wikipedia.org/wiki/Polygon) in order to further on implement  Jaccard and Haussdorf similarity  numbers.

addressing  #3115

### CODE REVIEW

File: src/Sets/Polygon.jl [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+algorithm on the line segments by default. This also removes redundant line segments.
+If the line segments are known to be sorted, the flag `apply_convex_hull=false` can
+be used to skip this preprocessing.
+
+### Examples
+
+```
+"""
+struct Polygon{N, VN<:AbstractVector{N}} <: AbstractPolygon{N}
+    line_segments::LineSegment{N, VN<:AbstractVector{N}}
```
Comment (mforets) [inline src/Sets/Polygon.jl:26]:
```suggestion
    edges::Vector{LineSegment{N, VN}}
```
we could in principle use any outer array type, but `Vector` is also fine.

File: src/Sets/Polygon.jl [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    # default constructor that applies a convex hull algorithm
+    function Polygon(line_segments::LineSegment{N, VN<:AbstractVector{N}};
+                      apply_convex_hull::Bool=true,
```
Comment (mforets) [inline src/Sets/Polygon.jl:30]: Why apply a convex hull?

File: src/Sets/Polygon.jl [hunk 3]
```diff
@@ -0,0 +1,38 @@
+export Polygon
```
Comment (mforets) [inline src/Sets/Polygon.jl:1]: Remember to include this file in LazySets.jl

File: src/Sets/Polygon.jl [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+To ensure this property, the constructor of `Polygon` runs a non-convex-hull
+algorithm on the line segments by default. This also removes redundant line segments.
+If the line segments are known to be sorted, the flag `apply_convex_hull=false` can
+be used to skip this preprocessing.
+
+### Examples
+
+```
+"""
+struct Polygon{N, VN<:AbstractVector{N}} <: AbstractPolygon{N}
```
Comment (mforets) [inline src/Sets/Polygon.jl:25]: Currently, the `AbstractPolygon` interface refers to convex polygons (seen as a subtype of polytopes) -- though we may want to review this. For the time being, you can just make it `<: LazySet{N}` directly.

**Review by mforets — COMMENTED** [review]
I had an initial look.

File: src/Sets/Polygon.jl (hunk 4 above)
Comment (mvanzulli) [reply src/Sets/Polygon.jl]: Done

File: src/Sets/Polygon.jl (hunk 3 above)
Comment (mvanzulli) [reply src/Sets/Polygon.jl]: Done

File: src/Sets/Polygon.jl (hunk 2 above)
Comment (mvanzulli) [reply src/Sets/Polygon.jl]: The name is not `convex-hull`, but I was considering to check redundant vertices and also counter-clock order, but maybe I'm wrong.

Comment (mforets) [reply src/Sets/Polygon.jl]: ok i guess we could check the order. it is preferable to apply if by default?

File: src/Sets/VPolygonNC.jl [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ [2.0, 0.0]
+ [1.0, 1.0]
+```
+
+```
+"""
+struct VPolygonNC{N, VN<:AbstractVector{N}} <: LazySet{N}
+    vertices::Vector{VN}
+
+    # default constructor that applies a convex hull algorithm
```
Comment (schillic) [inline src/Sets/VPolygonNC.jl:34]:
```suggestion
```

File: src/Sets/VPolygonNC.jl [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+julia> NCP.vertices
+5-element Vector{Vector{Float64}}:
+ [0.0, 0.0]
+ [0.0, 2.0]
+ [2.0, 2.0]
+ [2.0, 0.0]
+ [1.0, 1.0]
+```
+
+```
```
Comment (schillic) [inline src/Sets/VPolygonNC.jl:29]:
```suggestion
```

File: src/Sets/VPolygonNC.jl [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+Type that represents a non-convex polygon by its vertices.
+
+### Fields
+
+- `vertices` -- the list of vertices
+
+### Examples
+
+A non-convex polygon in vertex representation can be constructed by passing the list of
+vertices. For example, we can build a teeth:
```
Comment (schillic) [inline src/Sets/VPolygonNC.jl:15]:
```suggestion
vertices. For example, we can build a tooth:
```

**Review by schillic — COMMENTED** [review]
Maybe there should be a note about the order of the vertices in the documentation. For now we can allow both CW and CCW until we have a reason to choose, but you can also decide for one already.

Apart from that the PR looks good, thanks!
(@mforets can approve the build once ready.)

**Comment by mvanzulli** [comment]
> In this repo the docs job fails if there are missing docstrings. Please create a new file `VPolygonNC.md` under `docs/src/lib/sets`.

Done

**Comment by mforets** [comment]
The docs job passed but there are some doctests which error >> https://github.com/JuliaReach/LazySets.jl/actions/runs/3233891985/jobs/5296713250#step:6:1206

**Comment by mvanzulli** [comment]
> The docs job passed but there are some doctests which error >> https://github.com/JuliaReach/LazySets.jl/actions/runs/3233891985/jobs/5296713250#step:6:1206

At least locally it is solved 

```julia 
Test Summary:     | Pass  Total   Time
LazySets.doctests |    1      1  50.5s
 50.463812 seconds (77.29 M allocations: 5.399 GiB, 4.06% gc time, 58.11% compilation time: 1% of which was recompilation)
     Testing LazySets tests passed 
```

**Review by mforets — APPROVED** [review]
Tests passing! 

I have a few minor requests listed below, but we can make a new issue to address them.

- Add a function `vertices_list` similar to [this one](https://github.com/JuliaReach/LazySets.jl/blob/master/src/Sets/VPolygon.jl#L274)
- Add a function `tovrep` similar to [this one](https://github.com/JuliaReach/LazySets.jl/blob/master/src/Sets/VPolygon.jl#L199)
- Add a function `an_element` similar to [this one](https://github.com/JuliaReach/LazySets.jl/blob/master/src/Sets/VPolygon.jl#L410)
- Add functions [translate](https://github.com/JuliaReach/LazySets.jl/blob/master/src/Sets/VPolygon.jl#L596) / [translate!](https://github.com/JuliaReach/LazySets.jl/blob/master/src/Sets/VPolygon.jl#L624)
- Define [isoperationtype](https://github.com/JuliaReach/LazySets.jl/blob/master/src/Sets/VPolygon.jl#L84)
- Define the function `dim`, example [here](https://github.com/JuliaReach/LazySets.jl/blob/master/src/Interfaces/AbstractPolygon.jl#L47)
- Define `isconvextype`, example [here](https://github.com/JuliaReach/LazySets.jl/blob/master/src/Interfaces/AbstractPolygon.jl#L29)

**Comment by mforets** [comment]
> I have a few minor requests listed below, but we can make a new issue to address them.

See https://github.com/JuliaReach/LazySets.jl/issues/3124

**Review by schillic — APPROVED** [review]

**Comment by schillic** [comment]
Thanks again for your contribution @mvanzulli! We just published a new release and I added your name to the [release log](https://github.com/JuliaReach/LazySets.jl/releases/tag/v2.3.0). Let me know if you do not want to appear there.

**Pull request merged**